// react-blocks/lib/block.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var Block = (function (_Component) {
  _inherits(Block, _Component);

  function Block(props, context) {
    _classCallCheck(this, Block);

    _get(Object.getPrototypeOf(Block.prototype), 'constructor', this).call(this, props, context);
  }

  _createClass(Block, [{
    key: 'render',
    value: function render() {
      var props = this.props;
      var node = props.el || 'div';
      return _react2['default'].createElement(node, props);
    }
  }]);

  return Block;
})(_react.Component);

;

Block.PropTypes = {
  el: _react2['default'].PropTypes.string,
  block: _react2['default'].PropTypes.bool,
  hidden: _react2['default'].PropTypes.bool,
  invisible: _react2['default'].PropTypes.bool,
  relative: _react2['default'].PropTypes.bool,
  absolute: _react2['default'].PropTypes.bool,
  layout: _react2['default'].PropTypes.bool,
  inline: _react2['default'].PropTypes.bool,
  flex: _react2['default'].PropTypes.bool,
  vertical: _react2['default'].PropTypes.bool,
  horizontal: _react2['default'].PropTypes.bool,
  reverse: _react2['default'].PropTypes.bool,
  start: _react2['default'].PropTypes.bool,
  center: _react2['default'].PropTypes.bool,
  end: _react2['default'].PropTypes.bool,
  stretch: _react2['default'].PropTypes.bool,
  centered: _react2['default'].PropTypes.bool,
  selfStart: _react2['default'].PropTypes.bool,
  selfCenter: _react2['default'].PropTypes.bool,
  selfEnd: _react2['default'].PropTypes.bool,
  selfStretch: _react2['default'].PropTypes.bool,
  justifyStart: _react2['default'].PropTypes.bool,
  justifyCenter: _react2['default'].PropTypes.bool,
  justifyEnd: _react2['default'].PropTypes.bool,
  justifyBetween: _react2['default'].PropTypes.bool,
  justifyAround: _react2['default'].PropTypes.bool
};

exports['default'] = Block;
module.exports = exports['default'];
