// react-blocks/lib/layout.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _objectAssign = require('object-assign');

var _objectAssign2 = _interopRequireDefault(_objectAssign);

var common = undefined,
    layout = undefined,
    inline = undefined,
    flex = undefined,
    flexAuto = undefined,
    flexNone = undefined;
var horizontal = undefined,
    horizontalReverse = undefined,
    vertical = undefined,
    verticalReverse = undefined;
var alignStart = undefined,
    alignCenter = undefined,
    alignEnd = undefined,
    alignStretch = undefined;
var selfAlignStart = undefined,
    selfAlignCenter = undefined,
    selfAlignEnd = undefined,
    selfAlignStretch = undefined;
var justifyStart = undefined,
    justifyCenter = undefined,
    justifyEnd = undefined,
    justifyBetween = undefined,
    justifyAround = undefined;
var block = undefined,
    hidden = undefined,
    invisible = undefined;
var relative = undefined,
    absolute = undefined;
var centered = undefined;

// Defaults
common = {
  boxSizing: 'border-box',
  flexWrap: 'wrap'
};

layout = (0, _objectAssign2['default'])({}, common, {
  display: 'flex'
});

inline = (0, _objectAssign2['default'])({}, common, {
  display: 'inline-flex'
});

flex = {
  flexGrow: 1,
  flexShrink: 1,
  flexBasis: '0%'
};

flexAuto = (0, _objectAssign2['default'])({}, flex, {
  flexGrow: 1,
  flexShrink: 1,
  flexBasis: 'auto'
});

flexNone = {
  flexGrow: 0,
  flexShrink: 0,
  flexBasis: 'auto'
};

block = {
  display: 'block'
};

hidden = {
  display: 'none'
};

invisible = {
  visibility: 'hidden'
};

relative = {
  position: 'relative'
};

absolute = {
  position: 'absolute',
  top: 0,
  right: 0,
  bottom: 0,
  left: 0
};

horizontal = (0, _objectAssign2['default'])({}, layout, {
  flexDirection: 'row'
});

horizontalReverse = (0, _objectAssign2['default'])({}, layout, {
  flexDirection: 'row-reverse'
});

vertical = (0, _objectAssign2['default'])({}, layout, {
  flexDirection: 'column'
});

verticalReverse = (0, _objectAssign2['default'])({}, layout, {
  flexDirection: 'column-reverse'
});

alignStart = {
  alignItems: 'flex-start'
};

alignCenter = {
  alignItems: 'center'
};

alignEnd = {
  alignItems: 'flex-end'
};

alignStretch = {
  alignItems: 'stretch'
};

selfAlignStart = {
  alignSelf: 'flex-start'
};

selfAlignCenter = {
  alignSelf: 'center'
};

selfAlignEnd = {
  alignSelf: 'flex-end'
};

selfAlignStretch = {
  alignSelf: 'stretch'
};

justifyStart = {
  justifyContent: 'flex-start'
};

justifyCenter = {
  justifyContent: 'center'
};

justifyEnd = {
  justifyContent: 'flex-end'
};

justifyBetween = {
  justifyContent: 'space-between'
};

justifyAround = {
  justifyContent: 'space-around'
};

centered = (0, _objectAssign2['default'])({}, layout, alignCenter, justifyCenter);

exports['default'] = {
  block: block,
  hidden: hidden,
  invisible: invisible,
  centered: centered,
  relative: relative,
  absolute: absolute,
  layout: layout,
  inline: inline,
  flexAuto: flexAuto,
  flexNone: flexNone,
  horizontal: horizontal,
  horizontalReverse: horizontalReverse,
  vertical: vertical,
  verticalReverse: verticalReverse,
  alignStart: alignStart,
  alignCenter: alignCenter,
  alignEnd: alignEnd,
  alignStretch: alignStretch,
  selfAlignStart: selfAlignStart,
  selfAlignCenter: selfAlignCenter,
  selfAlignEnd: selfAlignEnd,
  selfAlignStretch: selfAlignStretch,
  justifyStart: justifyStart,
  justifyCenter: justifyCenter,
  justifyEnd: justifyEnd,
  justifyBetween: justifyBetween,
  justifyAround: justifyAround
};
module.exports = exports['default'];
